File: src/physics/PoolPhysics.ts
import { Vector2 } from './Vector2';
import { Ball, AimGuideData } from '../types';

export class PoolPhysics {
  static readonly TABLE_WIDTH = 1000;
  static readonly TABLE_HEIGHT = 500;

  static readonly MIN_X = 54;
  static readonly MAX_X = 946;
  static readonly MIN_Y = 54;
  static readonly MAX_Y = 446;

  static readonly BALL_RADIUS = 11.5;
  static readonly POCKET_RADIUS_CORNER = 22;
  static readonly POCKET_RADIUS_MIDDLE = 20;

  static readonly FRICTION = 0.988;
  static readonly CUSHION_RESTITUTION = 0.82;
  static readonly BALL_RESTITUTION = 0.92;
  static readonly VELOCITY_THRESHOLD = 0.08;

  static readonly POCKETS = [
    new Vector2(54, 54),       // Top-Left
    new Vector2(500, 48),      // Top-Middle
    new Vector2(946, 54),      // Top-Right
    new Vector2(54, 446),      // Bottom-Left
    new Vector2(500, 452),     // Bottom-Middle
    new Vector2(946, 446),     // Bottom-Right
  ];

  static readonly BALL_COLORS: Record<number, string> = {
    1: '#FBBF24', // Yellow
    2: '#3B82F6', // Blue
    3: '#EF4444', // Red
    4: '#8B5CF6', // Purple
    5: '#F97316', // Orange
    6: '#10B981', // Green
    7: '#991B1B', // Maroon / Dark Red
    8: '#0F172A', // Black (8-ball)
    9: '#FBBF24', // Stripe Yellow
    10: '#3B82F6', // Stripe Blue
    11: '#EF4444', // Stripe Red
    12: '#8B5CF6', // Stripe Purple
    13: '#F97316', // Stripe Orange
    14: '#10B981', // Stripe Green
    15: '#991B1B', // Stripe Maroon
  };

  static createInitialRack(): Ball[] {
    const balls: Ball[] = [];
    const r = PoolPhysics.BALL_RADIUS;

    // Cue Ball at head string
    balls.push({
      id: 0,
      number: 0,
      type: 'CUE',
      x: 280,
      y: 250,
      vx: 0,
      vy: 0,
      radius: r,
      baseColor: '#FFFFFF',
      isPotted: false,
      isSinking: false,
      pocketProgress: 0
    });

    // Standard 15-ball triangle rack at apex (x=700, y=250)
    const apexX = 680;
    const apexY = 250;
    const dx = r * 1.732; // cos(30 deg) * 2r
    const dy = r * 2.02;

    const rackNumbers = [
      1,
      2, 9,
      3, 8, 10,
      4, 14, 7, 11,
      5, 13, 15, 6, 12
    ];

    let index = 0;
    for (let row = 0; row < 5; row++) {
      const rowStartX = apexX + row * dx;
      const rowStartY = apexY - (row * dy) / 2;

      for (let col = 0; col <= row; col++) {
        const num = rackNumbers[index++];
        const type = num === 8 ? 'EIGHT' : (num <= 7 ? 'SOLID' : 'STRIPE');

        balls.push({
          id: num,
          number: num,
          type,
          x: rowStartX,
          y: rowStartY + col * dy,
          vx: 0,
          vy: 0,
          radius: r,
          baseColor: PoolPhysics.BALL_COLORS[num] || '#FFFFFF',
          isPotted: false,
          isSinking: false,
          pocketProgress: 0
        });
      }
    }

    return balls;
  }

  static stepSimulation(balls: Ball[], substeps: number = 4): boolean {
    let anyMoving = false;
    const dt = 1.0 / substeps;

    for (let step = 0; step < substeps; step++) {
      for (const ball of balls) {
        if (ball.isPotted) continue;

        if (ball.isSinking) {
          ball.pocketProgress += 0.08 * dt;
          if (ball.pocketProgress >= 1.0) {
            ball.isPotted = true;
            ball.isSinking = false;
            ball.vx = 0;
            ball.vy = 0;
          }
          anyMoving = true;
          continue;
        }

        const speedSq = ball.vx * ball.vx + ball.vy * ball.vy;
        if (speedSq > PoolPhysics.VELOCITY_THRESHOLD * PoolPhysics.VELOCITY_THRESHOLD) {
          anyMoving = true;
          ball.x += ball.vx * dt;
          ball.y += ball.vy * dt;

          const frictionCoeff = Math.pow(PoolPhysics.FRICTION, dt);
          ball.vx *= frictionCoeff;
          ball.vy *= frictionCoeff;
        } else {
          ball.vx = 0;
          ball.vy = 0;
        }

        // Pocket check
        for (const pocket of PoolPhysics.POCKETS) {
          const pdx = ball.x - pocket.x;
          const pdy = ball.y - pocket.y;
          const distSq = pdx * pdx + pdy * pdy;
          const pocketR = (pocket.x === 500 ? PoolPhysics.POCKET_RADIUS_MIDDLE : PoolPhysics.POCKET_RADIUS_CORNER) + 2;

          if (distSq < pocketR * pocketR) {
            ball.isSinking = true;
            ball.pocketProgress = 0;
            ball.vx *= 0.3;
            ball.vy *= 0.3;
            anyMoving = true;
            break;
          }
        }

        if (ball.isSinking) continue;

        // Cushion bounce
        const r = ball.radius;
        if (ball.x - r < PoolPhysics.MIN_X) {
          ball.x = PoolPhysics.MIN_X + r;
          ball.vx = -ball.vx * PoolPhysics.CUSHION_RESTITUTION;
        } else if (ball.x + r > PoolPhysics.MAX_X) {
          ball.x = PoolPhysics.MAX_X - r;
          ball.vx = -ball.vx * PoolPhysics.CUSHION_RESTITUTION;
        }

        if (ball.y - r < PoolPhysics.MIN_Y) {
          ball.y = PoolPhysics.MIN_Y + r;
          ball.vy = -ball.vy * PoolPhysics.CUSHION_RESTITUTION;
        } else if (ball.y + r > PoolPhysics.MAX_Y) {
          ball.y = PoolPhysics.MAX_Y - r;
          ball.vy = -ball.vy * PoolPhysics.CUSHION_RESTITUTION;
        }
      }

      // Ball-to-ball collisions
      for (let i = 0; i < balls.length; i++) {
        const b1 = balls[i];
        if (b1.isPotted || b1.isSinking) continue;

        for (let j = i + 1; j < balls.length; j++) {
          const b2 = balls[j];
          if (b2.isPotted || b2.isSinking) continue;

          const dx = b2.x - b1.x;
          const dy = b2.y - b1.y;
          const distSq = dx * dx + dy * dy;
          const minDist = b1.radius + b2.radius;

          if (distSq < minDist * minDist && distSq > 0.0001) {
            const dist = Math.sqrt(distSq);
            const nx = dx / dist;
            const ny = dy / dist;

            // Separate overlapping balls
            const overlap = (minDist - dist) * 0.5;
            b1.x -= nx * overlap;
            b1.y -= ny * overlap;
            b2.x += nx * overlap;
            b2.y += ny * overlap;

            // Elastic impulse
            const kx = b1.vx - b2.vx;
            const ky = b1.vy - b2.vy;
            const p = 2 * (nx * kx + ny * ky) / 2; // equal masses

            if (p > 0) {
              const impulse = p * PoolPhysics.BALL_RESTITUTION;
              b1.vx -= impulse * nx;
              b1.vy -= impulse * ny;
              b2.vx += impulse * nx;
              b2.vy += impulse * ny;
              anyMoving = true;
            }
          }
        }
      }
    }

    return anyMoving;
  }

  static calculateAimGuide(cueBall: Ball, balls: Ball[], aimAngle: Float64Array | number): AimGuideData {
    const angle = typeof aimAngle === 'number' ? aimAngle : aimAngle[0];
    const dir = Vector2.fromAngle(angle);
    const start = new Vector2(cueBall.x, cueBall.y);

    let closestDist = 1200;
    let hitBall: Ball | null = null;
    const r = PoolPhysics.BALL_RADIUS;

    for (const target of balls) {
      if (target.id === cueBall.id || target.isPotted || target.isSinking) continue;

      const toTarget = new Vector2(target.x - start.x, target.y - start.y);
      const proj = toTarget.dot(dir);
      if (proj <= 0) continue;

      const perpSq = toTarget.lengthSq() - proj * proj;
      const combinedR = r * 2;
      if (perpSq < combinedR * combinedR) {
        const d = proj - Math.sqrt(Math.max(0, combinedR * combinedR - perpSq));
        if (d < closestDist && d > 0) {
          closestDist = d;
          hitBall = target;
        }
      }
    }

    if (hitBall) {
      const ghostPos = start.add(dir.mul(closestDist));
      const targetPos = new Vector2(hitBall.x, hitBall.y);
      const normal = targetPos.sub(ghostPos).normalized();
      const targetTrajectory = normal.mul(70);
      const cueDeflection = dir.sub(normal.mul(dir.dot(normal))).normalized().mul(45);

      return {
        rayStartX: start.x,
        rayStartY: start.y,
        impactX: ghostPos.x,
        impactY: ghostPos.y,
        ghostCueX: ghostPos.x,
        ghostCueY: ghostPos.y,
        hasBallHit: true,
        targetBallX: targetPos.x + targetTrajectory.x,
        targetBallY: targetPos.y + targetTrajectory.y,
        cueDeflectX: ghostPos.x + cueDeflection.x,
        cueDeflectY: ghostPos.y + cueDeflection.y
      };
    }

    // Cushion raycast
    let rayLen = 900;
    if (dir.x > 0.001) rayLen = Math.min(rayLen, (PoolPhysics.MAX_X - r - start.x) / dir.x);
    else if (dir.x < -0.001) rayLen = Math.min(rayLen, (PoolPhysics.MIN_X + r - start.x) / dir.x);

    if (dir.y > 0.001) rayLen = Math.min(rayLen, (PoolPhysics.MAX_Y - r - start.y) / dir.y);
    else if (dir.y < -0.001) rayLen = Math.min(rayLen, (PoolPhysics.MIN_Y + r - start.y) / dir.y);

    const impact = start.add(dir.mul(Math.max(0, rayLen)));
    return {
      rayStartX: start.x,
      rayStartY: start.y,
      impactX: impact.x,
      impactY: impact.y,
      ghostCueX: impact.x,
      ghostCueY: impact.y,
      hasBallHit: false
    };
  }
}

File: app/src/main/java/com/example/game/PoolPhysics.kt
package com.example.game

import kotlin.math.acos
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

data class AimGuideData(
    val rayStart: Vector2,
    val impactPoint: Vector2,
    val ghostCuePosition: Vector2,
    val hasBallHit: Boolean,
    val hitBallId: Int? = null,
    val targetBallTrajectory: Vector2? = null,
    val cueBallDeflection: Vector2? = null
)

object PoolPhysics {
    const val TABLE_WIDTH = 1000f
    const val TABLE_HEIGHT = 500f

    const val MIN_X = 44f
    const val MAX_X = 956f
    const val MIN_Y = 44f
    const val MAX_Y = 456f

    const val BALL_RADIUS = 11.5f
    const val POCKET_RADIUS_CORNER = 27f
    const val POCKET_RADIUS_MIDDLE = 25f

    const val FRICTION = 0.990f
    const val RESTITUTION_BALL = 0.94f
    const val RESTITUTION_CUSHION = 0.85f

    val POCKETS = listOf(
        Vector2(40f, 40f),     // Top-Left
        Vector2(500f, 32f),    // Top-Middle
        Vector2(960f, 40f),    // Top-Right
        Vector2(40f, 460f),    // Bottom-Left
        Vector2(500f, 468f),   // Bottom-Middle
        Vector2(960f, 460f)    // Bottom-Right
    )

    fun createStandardBalls(): List<Ball> {
        val balls = mutableListOf<Ball>()

        // Ball 0: Cue ball at head line
        balls.add(
            Ball(
                id = 0,
                position = Vector2(280f, 250f),
                radius = BALL_RADIUS
            )
        )

        // Racked balls 1..15 in standard 8-ball triangle
        // Standard arrangement: 8 in center, corners one solid one stripe
        val rackOrder = listOf(
            1,            // Row 0
            9, 2,         // Row 1
            3, 8, 10,     // Row 2 (8-ball in center)
            11, 4, 12, 5, // Row 3
            6, 13, 7, 14, 15 // Row 4
        )

        val apexX = 700f
        val apexY = 250f
        val rowDx = BALL_RADIUS * sqrt(3f) + 0.3f
        val ballDy = BALL_RADIUS * 2f + 0.3f

        var orderIndex = 0
        for (row in 0..4) {
            val count = row + 1
            val rowX = apexX + row * rowDx
            val startY = apexY - (count - 1) * ballDy * 0.5f

            for (col in 0 until count) {
                if (orderIndex < rackOrder.size) {
                    val ballId = rackOrder[orderIndex]
                    val y = startY + col * ballDy
                    balls.add(
                        Ball(
                            id = ballId,
                            position = Vector2(rowX, y),
                            radius = BALL_RADIUS
                        )
                    )
                    orderIndex++
                }
            }
        }

        return balls
    }

    fun stepSimulation(
        balls: List<Ball>,
        subSteps: Int = 4,
        onBallHit: ((Ball, Ball) -> Unit)? = null,
        onBallPotted: ((Ball) -> Unit)? = null
    ) {
        val dt = 1.0f / subSteps

        repeat(subSteps) {
            // 1. Movement and pocket checks
            for (ball in balls) {
                if (ball.isPotted) continue

                if (ball.isSinking) {
                    // Ball is being sucked into a pocket
                    val target = ball.targetPocket ?: Vector2(0f, 0f)
                    val dir = (target - ball.position)
                    ball.position += dir * 0.25f
                    ball.pocketProgress += 0.08f
                    ball.velocity *= 0.5f
                    if (ball.pocketProgress >= 1f) {
                        ball.isPotted = true
                        ball.isSinking = false
                        ball.stop()
                        onBallPotted?.invoke(ball)
                    }
                    continue
                }

                // Normal movement
                ball.position += ball.velocity * dt
                ball.velocity *= FRICTION

                if (ball.velocity.lengthSquared() < 0.04f) {
                    ball.stop()
                }

                // Pocket entry check
                for (pocket in POCKETS) {
                    val pocketRadius = if (pocket.x == 500f) POCKET_RADIUS_MIDDLE else POCKET_RADIUS_CORNER
                    if (ball.position.distanceTo(pocket) < pocketRadius) {
                        ball.isSinking = true
                        ball.targetPocket = pocket
                        ball.pocketProgress = 0f
                        break
                    }
                }

                if (ball.isSinking) continue

                // Cushion bounces (ignoring pocket gaps)
                // Left cushion (between pockets)
                if (ball.position.x - ball.radius < MIN_X && ball.position.y in 66f..434f) {
                    ball.position = Vector2(MIN_X + ball.radius, ball.position.y)
                    ball.velocity = Vector2(-ball.velocity.x * RESTITUTION_CUSHION, ball.velocity.y)
                }
                // Right cushion
                if (ball.position.x + ball.radius > MAX_X && ball.position.y in 66f..434f) {
                    ball.position = Vector2(MAX_X - ball.radius, ball.position.y)
                    ball.velocity = Vector2(-ball.velocity.x * RESTITUTION_CUSHION, ball.velocity.y)
                }
                // Top cushions (left and right of center pocket)
                if (ball.position.y - ball.radius < MIN_Y) {
                    if (ball.position.x in 66f..474f || ball.position.x in 526f..934f) {
                        ball.position = Vector2(ball.position.x, MIN_Y + ball.radius)
                        ball.velocity = Vector2(ball.velocity.x, -ball.velocity.y * RESTITUTION_CUSHION)
                    }
                }
                // Bottom cushions (left and right of center pocket)
                if (ball.position.y + ball.radius > MAX_Y) {
                    if (ball.position.x in 66f..474f || ball.position.x in 526f..934f) {
                        ball.position = Vector2(ball.position.x, MAX_Y - ball.radius)
                        ball.velocity = Vector2(ball.velocity.x, -ball.velocity.y * RESTITUTION_CUSHION)
                    }
                }
            }

            // 2. Ball-to-ball collisions
            val activeBalls = balls.filter { !it.isPotted && !it.isSinking }
            for (i in activeBalls.indices) {
                for (j in i + 1 until activeBalls.size) {
                    val b1 = activeBalls[i]
                    val b2 = activeBalls[j]

                    val delta = b1.position - b2.position
                    val dist = delta.length()
                    val minDist = b1.radius + b2.radius

                    if (dist < minDist && dist > 0.0001f) {
                        val normal = delta / dist
                        val relVel = b1.velocity - b2.velocity
                        val velAlongNormal = relVel.dot(normal)

                        if (velAlongNormal < 0f) {
                            val impulse = -(1f + RESTITUTION_BALL) * velAlongNormal * 0.5f
                            b1.velocity += normal * impulse
                            b2.velocity -= normal * impulse

                            // Positional separation
                            val overlap = (minDist - dist) * 0.5f
                            b1.position += normal * overlap
                            b2.position -= normal * overlap

                            onBallHit?.invoke(b1, b2)
                        }
                    }
                }
            }
        }
    }

    fun calculateAimGuide(cueBall: Ball, aimAngleRadians: Float, allBalls: List<Ball>): AimGuideData {
        val rayStart = cueBall.position
        val rayDir = Vector2(cos(aimAngleRadians), sin(aimAngleRadians))

        var closestDistance = 1200f
        var hitBall: Ball? = null
        var ghostPos: Vector2 = rayStart + rayDir * 400f

        val objectBalls = allBalls.filter { !it.isCue && !it.isPotted && !it.isSinking }

        for (ball in objectBalls) {
            val toBall = ball.position - rayStart
            val projection = toBall.dot(rayDir)

            if (projection > 0f) {
                val perpDistSq = toBall.lengthSquared() - (projection * projection)
                val hitDistThresh = (BALL_RADIUS * 2f)
                val hitDistThreshSq = hitDistThresh * hitDistThresh

                if (perpDistSq <= hitDistThreshSq) {
                    val d = sqrt(hitDistThreshSq - perpDistSq)
                    val distanceToImpact = projection - d

                    if (distanceToImpact in 0f..closestDistance) {
                        closestDistance = distanceToImpact
                        hitBall = ball
                        ghostPos = rayStart + rayDir * distanceToImpact
                    }
                }
            }
        }

        if (hitBall != null) {
            val targetDir = (hitBall.position - ghostPos).normalized()
            val cueDeflection = (rayDir - targetDir * rayDir.dot(targetDir)).normalized()

            return AimGuideData(
                rayStart = rayStart,
                impactPoint = ghostPos,
                ghostCuePosition = ghostPos,
                hasBallHit = true,
                hitBallId = hitBall.id,
                targetBallTrajectory = targetDir * 120f,
                cueBallDeflection = cueDeflection * 60f
            )
        } else {
            // Check cushion intersection
            var cushionHitX = if (rayDir.x > 0) (MAX_X - BALL_RADIUS) else (MIN_X + BALL_RADIUS)
            var cushionHitY = if (rayDir.y > 0) (MAX_Y - BALL_RADIUS) else (MIN_Y + BALL_RADIUS)

            val tX = if (rayDir.x != 0f) (cushionHitX - rayStart.x) / rayDir.x else 9999f
            val tY = if (rayDir.y != 0f) (cushionHitY - rayStart.y) / rayDir.y else 9999f

            val t = if (tX in 0.1f..tY) tX else if (tY > 0.1f) tY else 500f
            val cushionImpact = rayStart + rayDir * t

            return AimGuideData(
                rayStart = rayStart,
                impactPoint = cushionImpact,
                ghostCuePosition = cushionImpact,
                hasBallHit = false,
                hitBallId = null,
                targetBallTrajectory = null,
                cueBallDeflection = null
            )
        }
    }
}
